### PR TITLE
Add support to build Host OS metapackages

### PR DESCRIPTION
**Pull request by olavphibm**
The OpenPOWER Host OS metapackages are divided into:
- open-power-host-os-release: install release file, with Host OS version
- open-power-host-os-all: install all Host OS packages
- open-power-host-os-*: install selected package group

All of the remaining open-power-host-os-* RPMs require the corresponding versions of the release RPM, so the user is able to link the metapackages installed with a Host OS version. They also install specific versions of the Host OS packages, but allow them to be overridden with custom versions after the installation.

### CODE REVIEW

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+
+%package all
+
+Summary: OpenPOWER Host OS whole experience
+
+Requires: %{name}-release = %{version}-%{release}
+Requires: %{name}-hypervisor-container = %{version}-%{release}
+Requires: %{name}-hypervisor-ui = %{version}-%{release}
+Requires: %{name}-hypervisor-virtualization = %{version}-%{release}
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: minor: AFAIK hypervisor in our case is QEMU + KVM. If they are installed by hypervisor-virtualization package, I do not understand why other packages have "hypervisor" in their names. IMHO "virtualization" would be a more accurate term for the virtualization tools installed by "hypervisor-container" and "hypervisor-ui". Also, I think that the rpm which installs the base set of packages should be more easily identified with a "base" in its name (or nothing). Therefore, my suggestion:

virtualization-container
virtualization-ui
virtualization-base/virtualization

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+Requires: %{name}-release = %{version}-%{release}
+
+Requires(post): kernel = 4.10.0-3.giteaaddb6.el7.centos
+Requires(post): libvirt = 2.2.0-6.gitddccbf6.el7.centos
+Requires(post): qemu = 15:2.8.0-1.gitbb80805.el7.centos
+
+
+%package ras
+
+Summary: OpenPOWER Host OS RAS packages
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: I'd mention RAS acronym meaning in the summary or description, IMHO it is not obvious to everyone. I myself had to google it.

File: open-power-host-os/open-power-host-os.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+Package:
+ sources: []
```
Comment (alanoe) [inline open-power-host-os/open-power-host-os.yaml]: side-note: I assume you have this [] here because our build code breaks if sources = None. IMHO we could fix that.

File: open-power-host-os/open-power-host-os.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+     - libvpd
+     - lshw
+     - lsvpd
+     - novnc
+     - ppc64-diag
+     - qemu
+     - servicelog
+     - sos
+     - systemtap
+     - wok
```
Comment (alanoe) [inline open-power-host-os/open-power-host-os.yaml]: side-note: when we infer install_dependencies from rpm specs, we can remove this list of all OP Host OS packages from here

**Comment by olavphibm** [comment]
A script in builds in necessary to keep those files updated. This script is not ready yet to deal with the package groups, so you'll have to wait a little for that PR.

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 1 above)
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Some points:

- none of the packages in 'container' depend on qemu/libvirt so I don't see why having them in the 'base' package. Maybe:
```
base  <-- kernel only
virtualization
virtualization-ui
container
```

- ui --> gui maybe? Qemu, libvirt and docker all have user interfaces.

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 5]
```diff
@@ -0,0 +1,145 @@
+Name: open-power-host-os
+Version: 7
```
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Why 7?

**Comment by fabianorosas** [comment]
This PR would address the issue described in: https://github.com/open-power-host-os/builds/issues/186#issuecomment-280097642

**Comment by mopsfelder** [comment]
What do you think about the following package set definition?

- `open-power-host-os-release`
- `open-power-host-os-all`: depends on -base, -kvm, -ras, -container, -management, -release
- `open-power-host-os-base`: depends on kernel, -release
- `open-power-host-os-kvm`: depends on qemu, libvirt, SLOF, -release
- `open-power-host-os-ras`: depends on librtas, lshw, hwdata, crash, powerpc-utls, systemtap, ppc64-diag, servicelog, libservicelog, -release
- `open-power-host-os-container`: depends on docker, docker-swarm, -release
- `open-power-host-os-management`: depends on kimchi, ginger, wok, -release

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 2 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Done

File: open-power-host-os/open-power-host-os.yaml (hunk 3 above)
Comment (olavphibm) [inline open-power-host-os/open-power-host-os.yaml]: Yes, I've fixed it in builds repo.

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 5 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: So it relates to CentOS and EPEL releases:
```
Requires: centos-release >= %{version}
Requires: epel-release >= %{version}
```
When we start supporting CentOS 8, we'll probably need two different releases (7 and 8), which will require a different set of packages with different versions.

**Comment by olavphibm** [comment]
I've applied @mopsfelder suggestion of package groups, merging `open-power-host-os-container-management` into `open-power-host-os-container`, because that appeared to be the most favored opinion among the team.

**Comment by olavphibm** [comment]
This depends on https://github.com/open-power-host-os/builds/pull/204

**Comment by olavphibm** [comment]
start build

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+Requires(post): gcc = 4.8.5-12.svn240558.el7.centos
+Requires(post): golang-github-russross-blackfriday = 1:1.2-6.git5f33e7b.el7.centos
+Requires(post): golang-github-shurcooL-sanitized_anchor_name = 1:0-1.git1dba4b3.el7.centos
+Requires(post): golang = 1.7.1-3.el7.centos
+
+
+%package base
+
+Summary: OpenPOWER Host OS hypervisor basic packages
```
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: s/hypervisor //

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+Requires(post): docker = 2:1.12.2-47.el7.centos
+Requires(post): docker-swarm = 1.1.0-1.gita0fd82b
+Requires(post): flannel = 0.5.5-1.gitcb8284f.el7.centos
+Requires(post): kubernetes = 1.2.0-0.21.git4a3f9c5.el7.centos
+
+
+%package kvm
+
+Summary: OpenPOWER Host OS hypervisor UI packages
```
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: s/UI //

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+Requires: %{name}-base = %{version}-%{release}
+
+Requires(post): SLOF = 20161019-4.gitba46a3f.el7.centos
+Requires(post): libvirt = 2.2.0-6.gitddccbf6.el7.centos
+Requires(post): qemu = 15:2.8.0-4.gita6ef3e6.el7.centos
+
+
+%package kvm-management
+
+Summary: OpenPOWER Host OS hypervisor UI packages
```
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: s/UI/management/

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 5 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: I'll change to 2.0 like you suggested.

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 6 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Done

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 7 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Done

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 8 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Done

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+%files all
+%files base
+%files container
+%files kvm
+%files kvm-management
+%files ras
+
+
+%changelog
+* Thu Mar 02 2017 Olav Philipp Henschel <olavph@linux.vnet.ibm.com> 7-1
```
Comment (fabianorosas) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: s/7/2.0/

**Comment by alanoe** [comment]
MO> open-power-host-os-kvm: depends on qemu, libvirt, SLOF, -base, -release
open-power-host-os-kvm-management: depends on kimchi, ginger, wok, -kvm, -base, -release

minor: all other metadata packages names contain functional roles (base, RAS, container), IMHO it'd be better to use "virtualization" or "virt" (functional role) instead of KVM (product name). Also KVM is just a virtualization support in the Linux kernel and does not depend on libvirt , libvirt is a wrapper API around several virtualization technologies. Therefore, it is odd that a package named KVM installs libvirt. Another reason to use the virtualization functional role in the package name.

File: open-power-host-os/CentOS/7/SOURCES/VERSION [hunk 10]
```diff
@@ -0,0 +1 @@
+../../../../VERSION
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/SOURCES/VERSION]: In short *and* long commit msg, packages => meta-packages 

minor: reading longer commit msg, it is ambiguous if the meta-packages  create the release file or if they allow the users to do it. Please rephrase it to eliminate the ambiguity. One possibility is using bullets for that, ie "The OpenPOWER Host OS packages:
- allow the users to easily install package groups
- create a release file (...)"

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+Summary: OpenPOWER Host OS release
+
+Requires: centos-release >= 7
+Requires: epel-release >= 7
+
+
+%package all
+
+Summary: OpenPOWER Host OS whole experience
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: minor: "whole experience" is completely unnecessary, seems a marketing description

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+Requires: %{name}-release = %{version}-%{release}
+
+Requires(post): kernel = 4.10.0-4.git8f27d36.el7.centos
+
+
+%package container
+
+Summary: OpenPOWER Host OS container packages
+
+Requires: %{name}-release = %{version}-%{release}
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Why require release package if base metapackage is required below and it, in turn, requires the release package?

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+%{summary}
+%description ras
+%{summary}
+
+
+%prep
+install -pm 644 %{SOURCE0} .
+
+%build
+echo "No build needed"
```
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: Do we need this echo?

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 9 above)
Comment (alanoe) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: I have no idea what "7-1". If it relates to the version (and it looks like by Fabiano's comment) imho something like "Initial 2.0 version. Builds packages <list of metapackages>." would be more useful.

**Comment by snowcrshd** [comment]
I'm depending on this PR getting merged in order to send the patches disabling SMT.

Are all the requested changes really necessary?

**Comment by alanoe** [comment]
LT> I'm depending on this PR getting merged in order to send the patches disabling SMT.

I don't get it. Reviewing is independent from merging. If any of your patches update this git repository and depend on this patch, base your patches on top of Olav's and submit the PR.

File: open-power-host-os/CentOS/7/open-power-host-os.spec (hunk 13 above)
Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: No. Removing.

**Comment by alanoe** [comment]
Pls check rpmlint output

File: open-power-host-os/CentOS/7/open-power-host-os.spec [hunk 14]
```diff
@@ -0,0 +1,157 @@
+%define milestone alpha
+%if "%{milestone}"
+%define milestone_tag .%{milestone}
+%endif
```
Comment (mopsfelder) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: I think the block

```rpm-spec
%define milestone alpha
%if "%{milestone}"
%define milestone_tag .%{milestone}
%endif
```

can be shortened to the following:

```rpm-spec
%define milestone alpha
%define milestone_tag %{?milestone:.%milestone:""}
```

Comment (olavphibm) [inline open-power-host-os/CentOS/7/open-power-host-os.spec]: I've tried and it does not work.
The `%{?milestone}` construction verifies whether the macro `milestone` is defined, so it will always evaluate to true and substitute.
We want to check whether `milestone` has a string value (alpha, beta) or not (%nil). The double quotes are needed for that.
